src/aleph/handlers/message_handler.py

```diff
@@ -316,6 +315,7 @@ async def process(
             session=session, pending_message=pending_message
         )
         content_handler = self.get_content_handler(message.type)
+        await content_handler.check_balance(session=session, message=message)
```

Comment:
This should definitely be called after `check_dependencies` (otherwise we might look for volumes that don't exist) and probably after `check_permissions` (we should reduce the amount of computation done before checking permissions).